dfdatetime/apfs_time.py

```diff
@@ -0,0 +1,121 @@
+
+  def __init__(self, timestamp=None):
+    """Initializes a APFS timestamp.
+
+    Args:
+      timestamp (Optional[int]): APFS timestamp.
+    """
+    super(APFSTime, self).__init__()
+    self._precision = definitions.PRECISION_1_NANOSECOND
+    self._timestamp = timestamp
+
+  @property
+  def timestamp(self):
+    """int: APFS timestamp or None if timestamp is not set."""
+    return self._timestamp
+
+  def _GetNormalizedTimestamp(self):
+    """Retrieves the normalized timestamp.
+
+    Returns:
+      decimal.Decimal: normalized timestamp, which contains the number of
+          seconds since January 1, 1970 00:00:00 and a fraction of second used
+          for increased precision, or None if the normalized timestamp cannot be
+          determined.
+    """
+    if self._normalized_timestamp is None:
+      if (self._timestamp is not None and self._timestamp >= 0 and
+          self._timestamp <= self._UINT64_MAX):
+        self._normalized_timestamp = (
+            decimal.Decimal(self._timestamp) /
+            definitions.NANOSECONDS_PER_SECOND)
+
+    return self._normalized_timestamp
+
+  def CopyFromDateTimeString(self, time_string):
+    """Copies a APFS timestamp from a date and time string.
+
+    Args:
+      time_string (str): date and time value formatted as:
+          YYYY-MM-DD hh:mm:ss.######[+-]##:##
+
+          Where # are numeric digits ranging from 0 to 9 and the seconds
+          fraction can be either 3 or 6 digits. The time of day, seconds
+          fraction and time zone offset are optional. The default time zone
+          is UTC.
+    """
+    date_time_values = self._CopyDateTimeFromString(time_string)
+
+    year = date_time_values.get('year', 0)
+    month = date_time_values.get('month', 0)
+    day_of_month = date_time_values.get('day_of_month', 0)
+    hours = date_time_values.get('hours', 0)
+    minutes = date_time_values.get('minutes', 0)
+    seconds = date_time_values.get('seconds', 0)
+    microseconds = date_time_values.get('microseconds', None)
+
+    timestamp = self._GetNumberOfSecondsFromElements(
+        year, month, day_of_month, hours, minutes, seconds)
+    timestamp *= definitions.NANOSECONDS_PER_SECOND
+
+    if microseconds:
+      nanoseconds = microseconds * definitions.MILLISECONDS_PER_SECOND
+      timestamp += nanoseconds
+
+    if year < 1970:
+      raise ValueError('Date time value not supported.')
+
+    # Maximum value for APFS time is 2554-07-21 23:34:33.709551615
+    if timestamp > self._UINT64_MAX:
+      raise ValueError('Date time value not supported.')
+
+    self._normalized_timestamp = None
+    self._timestamp = timestamp
+
+  def CopyToDateTimeString(self):
+    """Copies the APFS timestamp to a date and time string.
+
+    Returns:
+      str: date and time value formatted as: "YYYY-MM-DD hh:mm:ss.######" or
+          None if the timestamp is missing or invalid.
+    """
+    if (self._timestamp is None or self._timestamp < 0 or
```

Comment:
Is the APFS timestamp unsigned ? I would expect time_t and there for the timestamp to be signed.

Please change to check for INT64_MIN and INT64_MAX bounds instead

Reply:
paper pointed in #116 claims is UINT64, I haven't been able to confirm this claim yet.